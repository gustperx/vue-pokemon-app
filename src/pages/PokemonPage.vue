<template>
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
        <h1>¿Quien es este Pokemón?</h1>
        <PokemonPicture 
            :pokemon-id="pokemon.id" 
            :show-pokemon="showPokemon" />

        <PokemonOptions 
            :pokemons="pokemonArr" 
            @selection-pokemon="checkAnswer" />

        <template v-if="showAnswer">
            <h2>{{ message }}</h2>
            <button @click="newGame">
                Nuevo Juego
            </button>
        </template>
    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import { getPokemonOptions } from '@/helpers/getPokemonOptions'

export default {
    name: "PokemonPage",

    data() {
        return {
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },

    components: {
        PokemonPicture, 
        PokemonOptions
    },

    methods: {
        async mixedPokemonsArray() {
            this.pokemonArr = await getPokemonOptions()
            const rndInt = Math.floor(Math.random() * 4)
            this.pokemon = this.pokemonArr[rndInt]
        },
        checkAnswer(selectedId) {
            this.showPokemon = true
            this.showAnswer = true

            if (selectedId === this.pokemon.id) {
                this.message = `Correcto, es ${this.pokemon.name}`
            } else {
                this.message = `Oops, era ${this.pokemon.name}`
            }
        },
        newGame() {
            this.showPokemon = false
            this.showAnswer = false
            this.message = ''
            this.pokemon = null;
            this.pokemonArr = []
            this.mixedPokemonsArray()
        }
    },

    mounted() {
        this.mixedPokemonsArray()
    }
}
</script>
