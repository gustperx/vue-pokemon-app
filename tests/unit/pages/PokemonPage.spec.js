import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('debe de hacer match con el snapshot', () => {
        //
    })
    
    
    test('debe de llamar mixPokemonArray al montar', () => {
        
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixedPokemonsArray')
        const wrapper = shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })
    
    
    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[1],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })


    test('debe de mostrar los componentes PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[3],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const componentPicture = wrapper.findComponent(PokemonPicture);
        const componentOptions = wrapper.findComponent(PokemonOptions);

        // PokemonPicture debe existir
        expect(componentPicture.exists()).toBe(true)
        // PokemonOptions debe existir
        expect(componentOptions.exists()).toBe(true)
        
        // PokemonPicture attribute pokemonid === 5
        expect(componentPicture.attributes('pokemonid')).toBe('4')
        // PokemonOptions attribte pokemons toBe(true)
        expect(componentOptions.attributes('pokemons')).toBe('[object Object],[object Object],[object Object],[object Object]')
    })


    test('pruebas checkAnswer', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[3],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(4)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`Correcto, es ${pokemons[3].name}`)
    })


    test('pruebas checkAnswer - error', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[3],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[3].name}`)
    })
})