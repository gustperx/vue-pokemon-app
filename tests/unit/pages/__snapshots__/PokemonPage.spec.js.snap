// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quien es este Pokemón?</h1>
  <pokemon-picture-stub pokemonid="2" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
